2022/Q3/artifacts/PoS-tla/Staking.tla

```diff
@@ -113,104 +153,245 @@ Delegate(sender, validator, amount) ==
+                                         THEN 
+                                          LET min == Min(record.remaining, bond.amount) 
+                                          IN
+                                          [ remaining |-> record.remaining - min,
+                                            set |-> record.set \union {[id |-> record.id, amount |-> min, start |-> bond.epoch, end |-> e]},
+                                            id |-> record.id + 1 ]   
+                                         ELSE [ remaining |-> 0,
+                                                set |-> record.set,
+                                                id |-> record.id ]
+                                        IN ApaFoldSet(F, [ remaining |-> totalAmount, set |-> {}, id |-> idUnbonds], setBonds)
+
+\* Unbond `amount` tokens from a validator
 Unbond(sender, validator, amount) ==
     LET fail ==
         \/ amount < 0
-        \/ amount > delegated[UnbondingLength, sender, validator]
+        \/ TotalBonds(sender, validator) < amount
+        \/ validator \in frozenValidators[0]
     IN
     /\ lastTx' = [ id |-> nextTxId, tag |-> "unbond", fail |-> fail,
                    sender |-> sender, toAddr |-> validator, value |-> amount ]
     /\ failed' = (fail \/ failed)
+    /\ nextTxId' = nextTxId + 1
+    /\ txCounter' = txCounter + 1
+    /\ UNCHANGED <<epoch, balanceOf, posAccount, slashes, idSlashes, enqueuedSlashes, frozenValidators>>
     /\  IF fail
         THEN
-          UNCHANGED <<balanceOf, unbonded, delegated, bonded>>
+          UNCHANGED <<unbonded, idUnbonds, bonded, idBonds, totalDeltas>>
         ELSE
-          \* transaction succeeds
-          /\ unbonded' = [ unbonded EXCEPT ![UnbondingLength, sender] = @ + amount ]
-          /\ delegated' = [ delegated EXCEPT ![UnbondingLength, sender, validator] = @ - amount ]
-          /\ bonded' = [ bonded EXCEPT ![UnbondingLength, validator] = @ - amount]
-          /\ UNCHANGED  balanceOf
-          
-
-\* withdraw unbonded tokens
-Withdraw(sender) ==
-    LET fail == FALSE
+          LET newUnbonds == ComputedUnbonds(bonded[sender, validator], amount, epoch + UnbondingLength)
+          IN
+          /\ unbonded' = [ unbonded EXCEPT ![sender, validator] = @ \union newUnbonds.set ]
+          /\ idUnbonds' = newUnbonds.id
+          /\ bonded' = [ bonded EXCEPT ! [sender, validator] = @ \union {[ id |-> idBonds, epoch |-> epoch + UnbondingLength, amount |-> -1*amount ]}]
+          /\ idBonds' = idBonds + 1
+          /\ totalDeltas' = [ totalDeltas EXCEPT ![UnbondingLength*2, validator] = @ - amount]
+
+(*
+* For a given unbond, it computes the amount to be withdraw after applying a set of slashes.
+*)
+\* @type: (Set(SLASH), UNBOND) => Int;
+ApplySlashes(setSlashes, unbond) == 
+    LET 
+    \* @type: (Int, SLASH) => Int;
+    F(total, slash) == total - unbond.amount*slash.finalRate
+    IN ApaFoldSet(F, unbond.amount, setSlashes)
```

Comment:
Can't this become negative?

Reply:
It can if the validator misbehaves multiple times. Let's reason about it whenever we integrate the fix to #7

Reply:
OK